### ISSUE TITLE
home-manager integration?

### ISSUE THREAD

**Issue by Radvendii** [ryantm/agenix#50]
Is it possible to use agenix from inside a home-manager module? I don't see anything talking about this, so I'm guessing it's not possible, but I figured I'd check.

I realize I can *use* an agenix secret from inside a home-manager module by just accessing the `/run/secrets/blah` file, but I'd like to keep the file generation (i.e. `age.secrets.blah.file = "${self}/secrets/blah";`) close to the point of use, rather than off in it's own NixOS module.

**Comment by ryantm** [comment]
I don't think it will work by default. We'll need to update the module to work with it. This is definitely in scope for this project.

**Comment by Pacman99** [comment]
A few of us came to a similar conclusion a while ago: https://github.com/divnix/devos/discussions/319. Thanks for actually opening the issue here.

**Comment by Radvendii** [comment]
This seems to have been attempted: https://github.com/jordanisaacs/homeage

I wonder if it would be possible / desireable to merge that into `agenix`, or better to leave it separate.

**Comment by Gerschtli** [comment]
Because both project try to achieve the same result except different build environments, these should at some point be merged. Ping @jordanisaacs for your opinion on that.

**Comment by jordanisaacs** [comment]
The reason I wrote mine separately as I see my project being primarily a module for `home-manager`. My vision for doesn't involve a CLI (interested only in the module) and its on pace to become more intimately tied to home-manager. So while both are similar, I see them as fitting different use cases: agenix for the system and CLI, homeage for declarative use with home-manager

**Comment by jordanisaacs** [comment]
The reason I wrote mine separately as I see my project being primarily a module for home-manager. My vision for homeage doesn't involve a CLI (interested only in the module) and its on pace to become more intimately tied to home-manager. So while both are similar, I see them as fitting different use cases: agenix for the system and CLI, homeage for declarative use with home-manager

**Comment by Gerschtli** [comment]
I think it would create a more consistent experience if the modules for NixOS and home-manager work the same (from a UX point of view). I agree, that agenix currently only supports system configuration, but a home-manager integration is possible. Also I think there are many common logic parts or scripts (like module options, the shell script for linking/copying, the cleanup logic, etc.) that could be shared between NixOS and home-manager implementation. That would improve the quality of both project because the actual logic only needs to be written once.

And the CLI seems to be replaced by agenix-cli in the future anyway.

Opinions @jordanisaacs @ryantm?

**Comment by Nebucatnetzer** [comment]
Would be nice if agenix would work with home-manager as well.
Just for the sake of uniformity in the config.

**Comment by aouerf** [comment]
This can probably be closed because of #180 ?

**Comment by Nebucatnetzer** [comment]
I have seen that as well but how do you use it?On 25 Jun 2023, at 18:45, Ahmad Ouerfelli ***@***.***> wrote:﻿
This can probably be closed because of #180 ?

**Comment by sellout** [comment]
Yeah, I think it's reasonable to keep this open pending docs. I've used it already, so I can probably write it up today or tomorrow.

**Comment by Nebucatnetzer** [comment]
That would be great!On 25 Jun 2023, at 21:09, Greg Pfeil ***@***.***> wrote:﻿
Yeah, I think it's reasonable to keep this open pending docs. I've used it already, so I can probably write it up today or tomorrow.

**Comment by Brawl345** [comment]
any updates on this? Struggling to get it to work, tried adding `agenix.homeManagerModules.age` in my home-manager config to ìmports` but it keeps erroring out with "infinite recursion encountered" :/

**Comment by sellout** [comment]
Sorry, I apologize for not getting to this yet. I’ve only used it via flakes, where you do
```nix
    {
      homeConfigurations = {
        "user@host" = inputs.home-manager.lib.homeManagerConfiguration {
          modules = [
            inputs.agenix.homeManagerModules.age # ← the important bit
          ];
        };
      };
    };
```
for a non-flake config (like it sounds like you have, @Brawl345), I _think_ you should instead do something like
```nix
{
  imports = [ <agenix/modules/age-home.nix> ];
}
```
that assumes you’re using nix channels, but the instructions [here](https://github.com/ryantm/agenix#installation) should translate from NixOS to home-manager with a few substitutions:
- configuration.nix → home.nix
- age.nix -> age-home.nix
- nixosModules -> homeManagerModules

@Nebucatnetzer @Brawl345 – let me know if this short explanation solved your issues, so I know if my writeup will be correct.

**Comment by Nebucatnetzer** [comment]
I reckon with the examples by @sellout and @Brawl345 I should be able to integrate it.

**Comment by Nebucatnetzer** [comment]
Hm I'm a bit confused.
When I do `offlineimap -o` I get `cat: '$XDG_RUNTIME_DIR/agenix/personalEmailKey': No such file or directory` but when I run the command manually I see the password.
The config is basically the same just with the new module.
Here is a diff of the new config:
https://git.2li.ch/Nebucatnetzer/nixos/pulls/3/files

**Comment by ldicarlo** [comment]
Hi, and thank you for this awesome package. 
So I see the same error as @Nebucatnetzer.
It seems that 
`age.secrets.github_token.file = ./secrets/github_token.age;` results in 
`$(${pkgs.coreutils}/bin/cat ${config.age.secrets.github_token.path})` being evaluated to `/agenix/github_token`

Since it results in the log ` /agenix/github_token: No such file or directory` during switch.

(I use a flake).

`/agenix/` folder does not exists on my computer. What is the folder supposed to be when using Home Manager Agenix Module? 
Thank you

**Comment by ldicarlo** [comment]
I managed to do it. 
I think there are 2 problems:
- in my case it seems that `$XDG_RUNTIME_DIR` was not interpolated as my home folder and was empty, so the secrets would be looked for in `/agenix/` instead of `/ run/user/1000/agenix` but i did not fix neither understand rightly this problem.
- I think the user systemd service `agenix.service` was run only once and then not reran. Because once I did run it again then the secrets were updated and it worked.

Currently on a non-darwin system, this is what I have (and work): 

- (nowhere `agenix.nixosModules.default` ) 
- In my home configuration (`home-manager.users.luca`) : 
```

imports =
    [
      inputs.agenix.homeManagerModules.default
    ];
 age.secrets.secret1.file = ./secrets/sec1.age;

home.packages = [ 
    (pkgs.writeShellScriptBin "echo-secret" ''
      ${pkgs.coreutils}/bin/cat ${config.age.secrets.secret1.path}
    '')
];

```

secrets.nix with basic info like in the tutorial.

**Comment by Nebucatnetzer** [comment]
I can’t remember it from the top of my head but I think there is a way to force a restart of a service after a rebuild.
Might be a useful workaround for this.
I hadn’t had time yet to try a second time but I hope to be able to test it soon.

**Comment by shymega** [comment]
I'm having a similar issue with Home Manager+Agenix, and the `$XDG_RUNTIME_DIR`.

I'm basically storing my Taskwarrior sync CA/certificate/key, and credentials in Age-encrypted files.

My home manager configuration is fairly simple. The secrets can be decrypted, but the produced file for Taskwarrior keeps the `$XDG_RUNTIME_DIR` path in the config. This is invalid for Taskwarrior, and I'm a little unsure of how to proceed. I think this is something that could be extended in Agenix to have an output with the full, expanded path to the decrypted secret.

I really like Agenix, and I'm happy with it thus far. I just don't want to use solutions involving `cat` or `builtin.readFile` for the security risks involved there.

My configuration extract for Taskwarrior is listed below, as well as the produced configuration by Home Manager:


```nix
    taskwarrior = {
      enable = true;
      config = {
        confirmation = false;
        report = {
          minimal.filter = "status:pending";
          active.columns = [ "id" "start" "entry.age" "priority" "project" "due" "description" ];
          active.labels = [ "ID" "Started" "Age" "Priority" "Project" "Due" "Description" ];
        };
        taskd = {
          certificate = config.age.secrets.taskwarrior_sync_cert.path;
          key = config.age.secrets.taskwarrior_sync_key.path;
          ca = config.age.secrets.taskwarrior_sync_ca.path;
          server = "inthe.am:53589";
          credentials = config.age.secrets.taskwarrior_sync_cred.path;
        };
      };
    };
  };
```

```conf
data.location=/home/dzrodriguez/.local/share/task


confirmation=false
report.active.columns=id,start,entry.age,priority,project,due,description
report.active.labels=ID,Started,Age,Priority,Project,Due,Description
report.minimal.filter=status:pending
taskd.ca=$XDG_RUNTIME_DIR/agenix/taskwarrior_sync_ca
taskd.certificate=$XDG_RUNTIME_DIR/agenix/taskwarrior_sync_cert
taskd.credentials=$XDG_RUNTIME_DIR/agenix/taskwarrior_sync_cred
taskd.key=$XDG_RUNTIME_DIR/agenix/taskwarrior_sync_key
taskd.server=inthe.am:53589
```

**Comment by shymega** [comment]
> I can’t remember it from the top of my head but I think there is a way to force a restart of a service after a rebuild. Might be a useful workaround for this. I hadn’t had time yet to try a second time but I hope to be able to test it soon.

That would be really useful to know. If you do remember, maybe a PR to the documentation would be great!

**Comment by Nebucatnetzer** [comment]
Maybe extending `home.activation` helps. https://nix-community.github.io/home-manager/options.html#opt-home.activation
IIRC what I was using was the `userActivationScript` from NixOS.
Doesn't sound like the most optimal solution however.

**Comment by Deadbyrd12** [comment]
I solved the issue of restarting the service on my system by adding this line:
`systemd.user.startServices = "sd-switch";`
to my home-manager config. 

This changes the general behavior when it comes to starting/stopping services after switching generations. That means it might cause some weird side effects. The default is "suggest" which prints some suggestions as to which services should be restarted manually. "sd-switch" uses a 3rd-party app to start/stop the required services (requires a user dbus session). It seems this is going to become the default at some point. See the [systemd home-manager module](https://github.com/nix-community/home-manager/blob/408ba13188ff9ce309fa2bdd2f81287d79773b00/modules/systemd.nix#L162).

**Comment by jackwilsdon** [comment]
I'm currently using the following workaround for machines which do not have a dbus session (and therefore can't use systemd user units):

```nix
# In home-manager.users.your-name-here
home.activation.agenix = lib.hm.dag.entryAnywhere config.systemd.user.services.agenix.Service.ExecStart;
```

It's a bit of a hack as agenix doesn't expose the user activation script anywhere, but it appears to work well enough.

**Comment by jackwilsdon** [comment]
Turns out that the above only seems to work on switch and not on boot, as `XDG_RUNTIME_DIR` isn't set.

**Comment by kukovecz** [comment]
Hello,

I am really new in learning nix and using some base ubuntu (linux mint precisely) with nix as package manager and home manager to setup my config.

It was a little bit painful to setup agenix in this way, but it works, thanks for all the help in the documentation and people responses in this issue.

I want to leave the working configuration here if it might help anyone. 

I am using flakes and I've put this to my `flake.nix`:

```nix
homeConfigurations."username" = home-manager.lib.homeManagerConfiguration {
  inherit pkgs;

  modules = [
    # <other modules>
    agenix.homeManagerModules.age
    {
      age = {
        secrets = {
          "mysecret1" = {
            file = path/to/mysecret1.age;
            # owner = "username";  <-- This is not working
          };
        };
        secretsDir = "<home of username>/.agenix/agenix";
        secretsMountPoint = "<home of username>/.agenix/agenix.d";
        identityPaths = [ "<path/to/private/key/for/mysecret1.age>" ];
      };
    }
    # <other modules>
  ];
};
```

The `owner` setting is not working for the secret, despite stating in the [documentation](https://github.com/ryantm/agenix/tree/main?tab=readme-ov-file#agesecretsnameowner).

By not working, I mean
```shell
error: The option `age.secrets."mysecret1".owner' does not exist.
```

 I assume this is intentional (that option does not exist in [age-home.nix](https://github.com/ryantm/agenix/blob/main/modules/age-home.nix)), because as this being a home manager module, the owner will always be `username`.

**Comment by Yeshey** [comment]
I was trying to write a user service that would put my secrets where I wanted, apparently you need `systemd.user.services.<my_service>.Unit.After = [ "agenix.service" ];` for it to work, figured out by reading the sops-nix documentation, some similar documentation on agenix would be cool. with help fom @ldicarlo answer this is the service I got:

```nix
let
  inherit (pkgs.stdenv.hostPlatform) system; # for agenix pkg
  mystuff = pkgs.writeShellScriptBin "echo-secret" ''
        ${pkgs.coreutils}/bin/cat ${config.age.secrets.my_identity.path} > /home/yeshey/Downloads/ImOkay.txt
      '';
in
{
    home.packages = [
      inputs.agenix.packages.${system}.agenix
      mystuff # so now in the terminal running `echo-secret` echoes the secret
    ];

    systemd.user.services."test" = {
      Unit = {
        Description = "test";
        After = [ "agenix.service" ];
      };
      Service = {
        Type = "oneshot";
        ExecStart = "${mystuff}/bin/echo-secret";
      };
      Install.WantedBy = [ "default.target" ];
    };
}
```

**Comment by DerRockWolf** [comment]
> I solved the issue of restarting the service on my system by adding this line: `systemd.user.startServices = "sd-switch";` to my home-manager config.
> 
> This changes the general behavior when it comes to starting/stopping services after switching generations. That means it might cause some weird side effects. The default is "suggest" which prints some suggestions as to which services should be restarted manually. "sd-switch" uses a 3rd-party app to start/stop the required services (requires a user dbus session). It seems this is going to become the default at some point. See the [systemd home-manager module](https://github.com/nix-community/home-manager/blob/408ba13188ff9ce309fa2bdd2f81287d79773b00/modules/systemd.nix#L162).

FYI: since https://github.com/nix-community/home-manager/commit/b18f3ebc4029c22d437e3424014c8597a8b459a0 (2024-08-23) the value for `true` is `sd-switch` instead of `legacy`, and since https://github.com/nix-community/home-manager/commit/9a9fef316ad191b3086edda465e850af282de4e0 `sd-switch` is the default. 😄

**Comment by ccalhoun1999** [comment]
Is this at all functional? I've set `age.secretsDir = "${config.home.homeDirectory}/.agenix/agenix";` but no .agenix folder seems to be made anywhere, and using the path for `services.syncthing.passwordFile` doesn't amount to any password in syncthing (although it does build).

**Comment by mtalexan** [comment]
> The owner and group options don't seem to exist for me. File is working though.

The actual options available are only documented in the module code here: https://github.com/ryantm/agenix/blob/main/modules/age-home.nix#L157
With the `secrets` key using the `secretsType` defined further up, here: https://github.com/ryantm/agenix/blob/main/modules/age-home.nix#L99

You're right it doesn't look like the `owner` or `group` options are included, probably because the mounting and symlinking is running as a systemd user unit which is already restricted to the user and default group of the user who owns the unit.

**Comment by StarryFire** [comment]
I was able to solve this by explicitly setting the path to the private key that will be used to decrypt the secrets.

```secrets.nix
{
  globalPaths,
  config,
  ...
}:
{
  age = {
    identityPaths = [ "${config.home.homeDirectory}/.ssh/id" ];

    secrets = {
      "github" = {
        file = "${globalPaths.secrets}/github/github.age";
      };
    };
  };
}
```

and setting this in my home-manager configuration
```
home-manager.sharedModules = [ inputs.agenix.homeManagerModules.default ];
```

**Comment by c0d3h01** [comment]
> I was able to solve this by explicitly setting the path to the private key that will be used to decrypt the secrets.
> 
> {
>   globalPaths,
>   config,
>   ...
> }:
> {
>   age = {
>     identityPaths = [ "${config.home.homeDirectory}/.ssh/id" ];
> 
>     secrets = {
>       "github" = {
>         file = "${globalPaths.secrets}/github/github.age";
>       };
>     };
>   };
> }
> 
> and setting this in my home-manager configuration
> 
> home-manager.sharedModules = [ inputs.agenix.homeManagerModules.default ];

Can you share the Password File syntax how you using age keys on home manager
